### PR TITLE
Add Diet to OFF (WIP alpha 1)

### PR DESCRIPTION
**Pull request by dobriseb**
Add Diet (authorised, so-so and unauthorised ingredients) to OFF. 

## Description

The App will display authorised ingredients in green, so-so in orange, unauthaurised in red. Multi diets.
New Item Menu Your Diet(s) 
Click + to add a Diet,
Fill in the form,
Add another diet if desired then you can enable or disable a Diet.
Scan a product and take a look at the list of ingredients, your unauthorised ingredients will appear in red, the so-so ingredients in Orange and the authorised ones in Green.

 ## Screen-shots, if any
image
screenshot_wip_1_01
screenshot_wip_1_02
screenshot_wip_1_03
screenshot_wip_1_04

 
 ## Checklist
 
Make sure you've done all the following (You can delete the checklist before submitting)
 
 - [X] Apply the `AndroidStyle.xml` style template to your code in Android Studio.
 - [] Code is well documented
 - [ ] Include unit tests for new functionality
 - [ ] All user-visible strings are made translatable
 - [ ] Code passes Travis builds in your branch
 - [ ] If you have multiple commits please combine them into one commit by squashing them.
 - [X] Read and understood the contribution guidelines .

### CODE REVIEW

**Comment by teolemon** [comment]
```Could not find aapt2-proto.jar (com.android.tools.build:aapt2-proto:0.3.1).
Searched in the following locations:
    https://jcenter.bintray.com/com/android/tools/build/aapt2-proto/0.3.1/aapt2-proto-0.3.1.jar
```
Slight issues that are the same as described here: https://github.com/flutter/flutter/issues/23404

**Comment by teolemon** [comment]
```The specified Android SDK Build Tools version (27.0.3) is ignored, as it is below the minimum supported version (28.0.2) for Android Gradle Plugin 3.2.0.
Android SDK Build Tools 28.0.2 will be used.
To suppress this warning, remove "buildToolsVersion '27.0.3'" from your build.gradle file, as each version of the Android Gradle Plugin now has a default version of the build tools.
Update Build Tools version and sync project
Open File
```

**Comment by teolemon** [comment]
```2018-10-24 15:28:44.434 25533-25533/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: openfoodfacts.github.scrachx.openfood, PID: 25533
    java.lang.IllegalArgumentException: the bind value at index 1 is null
        at android.database.sqlite.SQLiteProgram.bindString(SQLiteProgram.java:164)
        at android.database.sqlite.SQLiteProgram.bindAllArgsAsStrings(SQLiteProgram.java:200)
        at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:49)
        at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1408)
        at android.database.sqlite.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1347)
        at org.greenrobot.greendao.database.StandardDatabase.rawQuery(StandardDatabase.java:32)
        at org.greenrobot.greendao.query.Query.list(Query.java:88)
        at org.greenrobot.greendao.query.QueryBuilder.list(QueryBuilder.java:427)
        at openfoodfacts.github.scrachx.openfood.repositories.DietRepository.getDietNameByDietTagAndLanguageCode(DietRepository.java:483)
        at openfoodfacts.github.scrachx.openfood.fragments.EditDietFragment.onCreateView(EditDietFragment.java:97)
        at android.support.v4.app.Fragment.performCreateView(Fragment.java:2346)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1428)
        at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1759)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1827)
        at android.support.v4.app.BackStackRecord.executeOps(BackStackRecord.java:797)
        at android.support.v4.app.FragmentManagerImpl.executeOps(FragmentManager.java:2596)
        at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2383)
        at android.support.v4.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2338)
        at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:2245)
        at android.support.v4.app.FragmentManagerImpl$1.run(FragmentManager.java:703)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)```

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -858,4 +858,19 @@
     <string name="legend_noael">above dose with observed adverse effect (NOAEL)</string>
     <string name="legend_adi">above Acceptable Daily Intake (ADI)</string>
     <!-- Used for additive exposure evaluation table -->
+    <!-- Used for diets AddOn -->
+    <string name="your_diets">Your diet(s)</string>
+    <string name="diet_name">Diet name</string>
+    <string name="diets_dialog_warning_title">Warning</string>
+    <string name="warning_diets_data">Your diet(s) data stay on your device. Nothing will be upload anywhere.</string>
```
Comment (teolemon) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="warning_diets_data">Your diet(s) data stays on your device. Nothing will be never be uploaded anywhere.</string>
```

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -858,4 +858,19 @@
...
+    <string name="EditDiet_Name">Name</string>
+    <string name="EditDiet_Description">Description</string>
+    <string name="EditDiet_Enabled_Disabled">Enabled/Disabled</string>
+    <string name="EditDiet_Authorised_Ingredients">Authorised ingredients</string>
+    <string name="EditDiet_Hint_Ingredient_Authorised">List the authorised ingredients that should appears in green. Example : Apple, Mushroom, Tomatoe... You can complete this list latter.</string>
```
Comment (teolemon) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="EditDiet_Hint_Ingredient_Authorised">List the authorised ingredients that should appear in green. Example : apple, mushroom, tomato… You can complete this list later.</string>
```

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -858,4 +858,19 @@
...
+    <string name="EditDiet_So_So_Ingredients">So so ingredients</string>
+    <string name="EditDiet_Hint_So_So_Ingredient">List the so so ingredients that should appears in orange. Example : Egg, Fish... You can complete this list latter.</string>
+    <string name="EditDiet_Unauthorised_Ingredients">Unauthorised ingredients</string>
+    <string name="EditDiet_Hint_Unauthorised_Ingredient">List the unauthorised ingredients that should appears in red. Example : Gluten, Milk, Palm oil, Sugar... You can complete this list latter.</string>
```
Comment (teolemon) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="EditDiet_Hint_Unauthorised_Ingredient">List the unauthorised ingredients that should appear in red. Example : Gluten, Milk, Palm oil, Sugar... You can complete this list later.</string>
```

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -858,4 +858,19 @@
...
+    <string name="diet_name">Diet name</string>
+    <string name="diets_dialog_warning_title">Warning</string>
+    <string name="warning_diets_data">Your diet(s) data stay on your device. Nothing will be upload anywhere.</string>
+    <string name="EditDiet_Name">Name</string>
+    <string name="EditDiet_Description">Description</string>
+    <string name="EditDiet_Enabled_Disabled">Enabled/Disabled</string>
+    <string name="EditDiet_Authorised_Ingredients">Authorised ingredients</string>
+    <string name="EditDiet_Hint_Ingredient_Authorised">List the authorised ingredients that should appears in green. Example : Apple, Mushroom, Tomatoe... You can complete this list latter.</string>
+    <string name="EditDiet_So_So_Ingredients">So so ingredients</string>
+    <string name="EditDiet_Hint_So_So_Ingredient">List the so so ingredients that should appears in orange. Example : Egg, Fish... You can complete this list latter.</string>
```
Comment (teolemon) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="EditDiet_Hint_So_So_Ingredient">List the problematic ingredients that should appear in orange. Example : Egg, Fish... You can complete this list later.</string>
```

File: app/src/main/res/values/strings.xml (hunk 2 above)
Comment (teolemon) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="EditDiet_Hint_Ingredient_Authorised">List the authorised ingredients that should appear in green. Example : apple, mushroom, tomato... You can complete this list later.</string>
```

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/EditDietFragment.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    Switch dietEnabled;
+    @BindView(R.id.ingredients_authorised)
+    EditText ingredientsAuthorised;
+    @BindView(R.id.ingredients_soso)
+    EditText ingredientsSoSo;
+    @BindView(R.id.ingredients_unauthorised)
+    EditText ingredientsUnauthorised;
+    @BindView(R.id.save_edits)
+    Button saveEdits;
+    //Récupération de la langue (théorique) de saisie :
```
Comment (teolemon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/EditDietFragment.java]: // Fetching of the (theoretical) language of input

Comment (teolemon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/EditDietFragment.java]:
```suggestion
    // Fetching of the (theoretical) language of input:
```

File: app/src/main/res/values/strings.xml (hunk 1 above)
Comment (dobriseb) [reply app/src/main/res/values/strings.xml]: Of course yes.

File: app/src/main/res/values/strings.xml (hunk 2 above)
Comment (dobriseb) [reply app/src/main/res/values/strings.xml]: The same, La prochaine fois je programme en français... quoique !

File: app/src/main/res/values/strings.xml (hunk 4 above)
Comment (dobriseb) [reply app/src/main/res/values/strings.xml]: You're the boss.

**Comment by teolemon** [comment]
We have an ingredient suggestion API. That would probably be useful to imput the list of ingredients.

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/EditDietFragment.java (hunk 5 above)
Comment (dobriseb) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/EditDietFragment.java]: Do you want me to translate my "Log" and comments ?

**Comment by dobriseb** [comment]
Some screenshots to be explicite :

The new Item Menu
screenshot_wip_1_01

A list of diets (toto is the only enable one)
screenshot_wip_1_02

A click on toto to edit this diet
screenshot_wip_1_03

The result in the list of ingredients of a scanned product
screenshot_wip_1_04

**Comment by teolemon** [comment]
the build is green :-) @dobriseb

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/DietsFragment.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                    .title(R.string.diets_dialog_warning_title)
+                    .content(R.string.warning_diets_data)
+                    .positiveText(R.string.ok_button)
+                    .show();
+            SharedPreferences.Editor editor = mSettings.edit();
+            editor.putBoolean("firstRunDiets", false);
+            editor.apply();
+        }
+
+        //Recherche des données à afficher dans le Recycler
```
Comment (teolemon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/DietsFragment.java]:
```suggestion
        // Search for data to display in the Recycler
```

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/DietsFragment.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        setHasOptionsMenu(true);
+        return createView(inflater, container, R.layout.fragment_diets);
+    }
+
+    public void onViewCreated(@NonNull final View view, @Nullable Bundle savedInstanceState) {
+        super.onViewCreated(view, savedInstanceState);
+        //Gestion du message d'avertissement, les données restent sur l'appareil.
```
Comment (teolemon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/DietsFragment.java]:
```suggestion
        // Management of the warning message, the data remains on the device.
```

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/DietsFragment.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+            diet1.setTag("en:Vegetarian");
+            diet1.setEnabled(true);
+            dietList.add(diet1);
+            Diet diet2 = new Diet();
+            diet2.setTag("en:Vegan");
+            diet2.setEnabled(false);
+            dietList.add(diet2);
+        }
+
+        //Activation du recycler avec l'adapter qui va bien
```
Comment (teolemon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/DietsFragment.java]:
```suggestion
        // Enabling the recycler with the proper adapter
```

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/DietsFragment.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        }));
+    }
+
+    /**
+     * Ajout d'un régime.
+     * Add a diet.
+     */
+    @OnClick(R.id.fab)
+    void openFragmentAddDiet () {
+        Log.i("INFO", "Début de onClick sur fab de FragmentDiets");
```
Comment (teolemon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/DietsFragment.java]:
```suggestion
        Log.i("INFO", "Starting onClick on fab from FragmentDiets");
```

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/EditDietFragment.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+import openfoodfacts.github.scrachx.openfood.models.Diet;
+import openfoodfacts.github.scrachx.openfood.models.DietName;
+import openfoodfacts.github.scrachx.openfood.repositories.DietRepository;
+import openfoodfacts.github.scrachx.openfood.repositories.IDietRepository;
+
+public class EditDietFragment extends Fragment {
+    // TODO: Rename parameter arguments, choose names that match
+    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
+    private static final String ARG_DIET_NAME = "dietName";
+    //Récupération des éléments du fragment de saisie d'une Diet
```
Comment (teolemon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/EditDietFragment.java]:
```suggestion
    // Retrieving elements of the input fragment of a Diet
```

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/DatabaseHelper.java [hunk 11]
```diff
@@ -133,6 +133,15 @@ private void upgrade(Database db, int migrateVersion) {
                 }
                 break;
             }
+            case 11: {
+                //Log.i("INFO", "Upgrade to version 11 ! ");
```
Comment (teolemon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/models/DatabaseHelper.java]:
```suggestion
                //Log.i("INFO", "Upgrading to version 11");
```

File: app/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -858,4 +858,19 @@
     <string name="legend_noael">above dose with observed adverse effect (NOAEL)</string>
     <string name="legend_adi">above Acceptable Daily Intake (ADI)</string>
     <!-- Used for additive exposure evaluation table -->
+    <!-- Used for diets AddOn -->
+    <string name="your_diets">Your diet(s)</string>
+    <string name="diet_name">Diet name</string>
+    <string name="diets_dialog_warning_title">Warning</string>
+    <string name="warning_diets_data">Your diet(s) data stays on your device. Nothing will be upload anywhere.</string>
```
Comment (teolemon) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="warning_diets_data">Your diet data always stays on your device. Nothing will be uploaded anywhere.</string>
```

File: app/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -858,4 +858,19 @@
...
+    <string name="EditDiet_Name">Name</string>
+    <string name="EditDiet_Description">Description</string>
+    <string name="EditDiet_Enabled_Disabled">Enabled/Disabled</string>
+    <string name="EditDiet_Authorised_Ingredients">Authorised ingredients</string>
+    <string name="EditDiet_Hint_Ingredient_Authorised">List the authorised ingredients that should appear in green. Example : Apple, Mushroom, Tomatoe... You can complete this list latter.</string>
```
Comment (teolemon) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="EditDiet_Hint_Ingredient_Authorised">List the authorized ingredients that should appear in green. Example : apple, mushroom, tomato… You can complete this list later.</string>
```

File: app/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -858,4 +858,19 @@
...
+    <string name="EditDiet_So_So_Ingredients">So so ingredients</string>
+    <string name="EditDiet_Hint_So_So_Ingredient">List the problematic ingredients that should appear in orange. Example : Egg, Fish... You can complete this list latter.</string>
```
Comment (teolemon) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="EditDiet_Hint_So_So_Ingredient">List the problematic ingredients that should appear in orange. Example : egg, fish… You can complete this list later.</string>
```

File: app/src/main/res/values/strings.xml [hunk 15]
```diff
@@ -858,4 +858,19 @@
...
+    <string name="EditDiet_Unauthorised_Ingredients">Unauthorised ingredients</string>
+    <string name="EditDiet_Hint_Unauthorised_Ingredient">List the unauthorised ingredients that should appear in red. Example : Gluten, Milk, Palm oil, Sugar... You can complete this list latter.</string>
```
Comment (teolemon) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="EditDiet_Hint_Unauthorised_Ingredient">List the unauthorized ingredients that should appear in red. Example : gluten, milk, palm oil, sugar... You can complete this list later.</string>
```

File: app/src/main/res/values/strings.xml [hunk 16]
```diff
@@ -858,4 +858,19 @@
...
     <!-- Used for additive exposure evaluation table -->
+    <!-- Used for diets AddOn -->
+    <string name="your_diets">Your diet(s)</string>
+    <string name="diet_name">Diet name</string>
+    <string name="diets_dialog_warning_title">Warning</string>
+    <string name="warning_diets_data">Your diet(s) data stays on your device. Nothing will be upload anywhere.</string>
+    <string name="EditDiet_Name">Name</string>
+    <string name="EditDiet_Description">Description</string>
+    <string name="EditDiet_Enabled_Disabled">Enabled/Disabled</string>
+    <string name="EditDiet_Authorised_Ingredients">Authorised ingredients</string>
```
Comment (teolemon) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="EditDiet_Authorised_Ingredients">Authorized ingredients</string>
```

File: app/src/main/res/values/strings.xml [hunk 17]
```diff
@@ -858,4 +858,19 @@
...
+    <string name="EditDiet_Hint_Ingredient_Authorised">List the authorised ingredients that should appear in green. Example : Apple, Mushroom, Tomatoe... You can complete this list latter.</string>
+    <string name="EditDiet_So_So_Ingredients">So so ingredients</string>
+    <string name="EditDiet_Hint_So_So_Ingredient">List the problematic ingredients that should appear in orange. Example : Egg, Fish... You can complete this list latter.</string>
+    <string name="EditDiet_Unauthorised_Ingredients">Unauthorised ingredients</string>
```
Comment (teolemon) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="EditDiet_Unauthorised_Ingredients">Unauthorized ingredients</string>
```

**Comment by dobriseb** [comment]
Close because there no more automatic updates from the fork...

**Comment by dobriseb** [comment]
Closed because there was no updates from the fork.
A new Pull Request is open.

**Pull request closed without merging**